From 4397a922f1c78b148c84c7dcc4a00774af2698df Mon Sep 17 00:00:00 2001
From: iuncuim <iuncuim@gmail.com>
Date: Mon, 5 Aug 2019 11:18:05 +0300
Subject: [PATCH] yjytj

---
 include/configs/somlabs_visionsom_6ull.h | 12 ++++++------
 1 file changed, 6 insertions(+), 6 deletions(-)

diff --git a/include/configs/somlabs_visionsom_6ull.h b/include/configs/somlabs_visionsom_6ull.h
index a26d413..1d3d6f8 100644
--- a/include/configs/somlabs_visionsom_6ull.h
+++ b/include/configs/somlabs_visionsom_6ull.h
@@ -60,7 +60,7 @@
 	"fdt_high=0xffffffff\0"	  \
 	"console=ttymxc0\0" \
 	"script=boot.scr\0" \
-	"image=zImage\0" \
+	"image=uImage\0" \
 	"splashimage=0x80000000\0" \
 	"splashfile=/boot/splash.bmp\0" \
 	"setrootnand=setenv rootspec root=ubi0:rootfs ubi.mtd=ubi rootfstype=ubifs "MFG_NAND_PARTITION"\0" \
@@ -69,17 +69,17 @@
 	            "setenv loadfdt ubifsload ${fdt_addr} /boot/${fdt_file};\0"\
 	"mmcdev=0\0" \
 	"mmcpart=1\0" \
-	"mmcroot=/dev/mmcblk1p1 rootwait rw\0" \
+	"mmcroot=/dev/mmcblk1p2 rootwait rw\0" \
 	"setrootmmc=setenv rootspec root=${mmcroot}\0" \
 	"setbootscriptmmc=setenv loadbootscript ext4load mmc ${mmcdev}:${mmcpart} ${loadaddr} /boot/${script};\0" \
-	"setloadmmc=setenv loadimage ext4load mmc ${mmcdev}:${mmcpart} ${loadaddr} /boot/${image}; " \
-	           "setenv loadfdt ext4load mmc ${mmcdev}:${mmcpart} ${fdt_addr} /boot/${fdt_file};\0" \
+	"setloadmmc=setenv loadimage fatload mmc ${mmcdev}:${mmcpart} ${loadaddr} ${image}; " \
+	           "setenv loadfdt fatload mmc ${mmcdev}:${mmcpart} ${fdt_addr} ${fdt_file};\0" \
 	"setbootargs=setenv bootargs console=${console},${baudrate} " \
 		BOOTARGS_CMA_SIZE \
 		"${rootspec}\0" \
 	"execbootscript=echo Running bootscript...; " \
 		"source\0" \
-	"setfdtfile=setenv fdt_file somlabs-${board}${fdt_suffix}.dtb\0" \
+	"setfdtfile=setenv fdt_file dtb\0" \
 	"checkbootdev=if test ${bootdev} = nand; then " \
 		"nand device 0; ubi part ubi; ubifsmount ubi0:rootfs; " \
 		"run setbootscriptnand; " \
@@ -100,7 +100,7 @@
 		"else " \
 			"if run loadimage; then " \
 				"run setbootargs; " \
-				"bootz ${loadaddr} - ${fdt_addr}; " \
+				"bootm ${loadaddr} - ${fdt_addr}; " \
 		   "fi; " \
 	   "fi"
 
-- 
2.7.4

