--- a/arch/arm64/boot/dts/rockchip/rk3328-doornet1.dts
+++ b/arch/arm64/boot/dts/rockchip/rk3328-doornet1.dts
@@ -20,6 +20,8 @@
 		led-failsafe = &sys_led;
 		led-running = &sys_led;
 		led-upgrade = &sys_led;
+		// mmc1 = &sdmmc;
+		// mmc0 = &emmc;
 	};
 
 	chosen {
@@ -453,6 +455,20 @@
 	status = "okay";
 };
 
+&emmc {
+	bus-width = <8>;
+	cap-mmc-highspeed;
+	max-frequency = <150000000>;
+	mmc-ddr-1_8v;
+	mmc-hs200-1_8v;
+	non-removable;
+	pinctrl-names = "default";
+	pinctrl-0 = <&emmc_clk &emmc_cmd &emmc_bus8>;
+	vmmc-supply = <&vcc_io_33>;
+	vqmmc-supply = <&vcc18_emmc>;
+	status = "okay";
+};
+
 &tsadc {
 	rockchip,hw-tshut-mode = <0>;
 	rockchip,hw-tshut-polarity = <0>;
@@ -504,3 +520,4 @@
 	};
 };
 
+
