Index: linux-5.4.105/arch/arm64/boot/dts/rockchip/rk3399.dtsi
===================================================================
--- linux-5.4.105.orig/arch/arm64/boot/dts/rockchip/rk3399.dtsi
+++ linux-5.4.105/arch/arm64/boot/dts/rockchip/rk3399.dtsi
@@ -782,7 +782,7 @@
 					type = "passive";
 				};
 				cpu_alert1: cpu_alert1 {
-					temperature = <75000>;
+					temperature = <85000>;
 					hysteresis = <2000>;
 					type = "passive";
 				};
